litex/soc/cores/i2s.py

```diff
@@ -413,53 +538,85 @@ def __init__(self, pads, fifo_depth=256):
             txi2s.act("LEFT",
                 If(~self.tx_ctl.fields.enable,
                     NextState("IDLE")
                 ).Else(
-                    NextValue(tx_pin, tx_buf[31]),
+                    NextValue(tx_pin, tx_buf[sample_width - 1]),
                     NextValue(tx_buf, Cat(0, tx_buf[:-1])),
                     NextValue(tx_cnt, tx_cnt - 1),
                     NextState("LEFT_WAIT")
                 )
             )
-            txi2s.act("LEFT_WAIT",
-                If(~self.tx_ctl.fields.enable,
-                    NextState("IDLE")
-                ).Else(
-                    If(falling_edge,
-                        If((tx_cnt == 0) & sync_pin,
-                            NextValue(tx_cnt, 16),
-                            NextState("RIGHT")
-                        ).Elif(tx_cnt > 0,
-                            NextState("LEFT")
+            if concatenate_channels:
+                txi2s.act("LEFT_WAIT",
+                    If(~self.tx_ctl.fields.enable,
+                        NextState("IDLE")
+                    ).Else(
+                        If(falling_edge,
+                            If((tx_cnt == 0),
+                                If((sync_pin if frame_format == I2S_FORMAT.I2S_STANDARD else ~sync_pin),
+                                    NextValue(tx_cnt, sample_width),
+                                    NextState("RIGHT"),
+                                ).Else(
+                                    NextState("LEFT_WAIT"),
+                                )
+                            ).Elif(tx_cnt > 0,
+                                NextState("LEFT"),
+                            )
                         )
                     )
                 )
-            )
+            else:
+                txi2s.act("LEFT_WAIT",
+                    If(~self.tx_ctl.fields.enable,
+                        NextState("IDLE")
+                    ).Else(
+                        If(falling_edge,
+                            If((tx_cnt == 0),
+                                If((sync_pin if frame_format == I2S_FORMAT.I2S_STANDARD else ~sync_pin),
+                                    NextValue(tx_cnt, sample_width),
+                                    NextState("RIGHT"),
+                                    NextValue(tx_buf, Cat(tx_rd_d,offset)),
+                                    tx_rden.eq(1),
+                                ).Else(
+                                    NextState("LEFT_WAIT"),
+                                )
+                            ).Elif(tx_cnt > 0,
+                                NextState("LEFT"),
+                            )
+                        )
+                    )
+                )
+
             txi2s.act("RIGHT",
                 If(~self.tx_ctl.fields.enable,
                     NextState("IDLE")
                 ).Else(
-                    NextValue(tx_pin, tx_buf[31]),
+                    NextValue(tx_pin, tx_buf[sample_width-1]),
```

Comment:
same problem here of indexing into the wrong part of the buffer for the case of concatenate_channels

Reply:
fixed, as described above